tests/pubsub/testgossipsubmessagecache.nim

```diff
@@ -0,0 +1,312 @@
+{.used.}
+
+import std/[sequtils]
+import stew/byteutils
+import utils
+import ../../libp2p/protocols/pubsub/[gossipsub, mcache, peertable, floodsub]
+import ../../libp2p/protocols/pubsub/rpc/[messages, message]
+import ../helpers
+
+suite "GossipSub Message Cache":
+  teardown:
+    checkTrackers()
+
+  const
+    timeout = 1.seconds
+    interval = 50.milliseconds
+
+  asyncTest "Received messages are added to the message cache":
+    const
+      numberOfNodes = 2
+      topic = "foobar"
+    let nodes = generateNodes(numberOfNodes, gossip = true).toGossipSub()
+
+    startNodesAndDeferStop(nodes)
+
+    await connectNodesStar(nodes)
+    subscribeAllNodes(nodes, topic, voidTopicHandler)
+    await waitForHeartbeat()
+
+    # When Node0 publishes a message to the topic
+    tryPublish await nodes[0].publish("foobar", "Hello!".toBytes()), 1
+
+    # Then Node1 receives the message and saves it in the cache 
+    checkUntilCustomTimeout(timeout, interval):
+      nodes[1].mcache.window(topic).toSeq().len == 1
+
+  asyncTest "Message cache history shifts on heartbeat and is cleared on shift":
+    const
+      numberOfNodes = 2
+      topic = "foobar"
+      historyGossip = 3 # mcache window
+      historyLength = 5
+    let nodes = generateNodes(
+        numberOfNodes,
+        gossip = true,
+        historyGossip = historyGossip,
+        historyLength = historyLength,
+      )
+      .toGossipSub()
+
+    startNodesAndDeferStop(nodes)
+
+    await connectNodesStar(nodes)
+    subscribeAllNodes(nodes, topic, voidTopicHandler)
+    await waitForHeartbeat()
+
+    # When Node0 publishes a message to the topic
+    tryPublish await nodes[0].publish("foobar", "Hello!".toBytes()), 1
+
+    # Then Node1 receives the message and saves it in the cache 
+    checkUntilCustomTimeout(timeout, interval):
+      nodes[1].mcache.window(topic).toSeq().len == 1
```

Comment:
`nodes[1].mcache.window(topic).toSeq().len == 1`
does not need indirection with `toSeq()` 
`nodes[1].mcache.window(topic).len == 1` 
should work.
please update other places in this PR as well.